backends/events/default/default-events.cpp

```diff
@@ -86,163 +83,181 @@ void DefaultEventManager::init() {
-			_buttonState &= ~LBUTTON;
-			break;
+void DefaultEventManager::handleKeyRepeat() {
+	uint32 time = g_system->getMillis(true);
 
-		case Common::EVENT_RBUTTONDOWN:
-			_mousePos = event.mouse;
-			_buttonState |= RBUTTON;
-			break;
+	if (!_eventQueue.empty()) {
+		// Peek in the event queue
+		const Common::Event &nextEvent = _eventQueue.front();
 
-		case Common::EVENT_RBUTTONUP:
-			_mousePos = event.mouse;
-			_buttonState &= ~RBUTTON;
+		switch (nextEvent.type) {
+		case Common::EVENT_KEYDOWN:
+			// init continuous event stream
+			_currentKeyDown = nextEvent.kbd;
+			_keyRepeatTime = time + kKeyRepeatInitialDelay;
 			break;
 
-		case Common::EVENT_MAINMENU:
-			if (g_engine && !g_engine->isPaused())
-				g_engine->openMainMenuDialog();
-
-			if (_shouldQuit)
-				event.type = Common::EVENT_QUIT;
-			else if (_shouldRTL)
-				event.type = Common::EVENT_RTL;
-			break;
-#ifdef ENABLE_VKEYBD
-		case Common::EVENT_VIRTUAL_KEYBOARD:
-			if (_vk->isDisplaying()) {
-				_vk->close(true);
-			} else {
-				if (g_engine)
-					g_engine->pauseEngine(true);
-				_vk->show();
-				if (g_engine)
-					g_engine->pauseEngine(false);
-				result = false;
-			}
-			break;
-#endif
-#ifdef ENABLE_KEYMAPPER
-		case Common::EVENT_KEYMAPPER_REMAP:
-			if (!_remap) {
-				_remap = true;
-				Common::RemapDialog _remapDialog;
-				if (g_engine)
-					g_engine->pauseEngine(true);
-				_remapDialog.runModal();
-				if (g_engine)
-					g_engine->pauseEngine(false);
-				_remap = false;
+		case Common::EVENT_KEYUP:
+			if (nextEvent.kbd.keycode == _currentKeyDown.keycode) {
+				// Only stop firing events if it's the current key
+				_currentKeyDown.keycode = Common::KEYCODE_INVALID;
 			}
 			break;
-#endif
-		case Common::EVENT_RTL:
-			if (ConfMan.getBool("confirm_exit")) {
-				if (g_engine)
-					g_engine->pauseEngine(true);
-				GUI::MessageDialog alert(_("Do you really want to return to the Launcher?"), _("Launcher"), _("Cancel"));
-				result = _shouldRTL = (alert.runModal() == GUI::kMessageOK);
-				if (g_engine)
-					g_engine->pauseEngine(false);
-			} else
-				_shouldRTL = true;
-			break;
-
-		case Common::EVENT_MUTE:
-			if (g_engine)
-				g_engine->flipMute();
-			break;
-
-		case Common::EVENT_QUIT:
-			if (ConfMan.getBool("confirm_exit")) {
-				if (_confirmExitDialogActive) {
-					result = false;
-					break;
-				}
-				_confirmExitDialogActive = true;
-				if (g_engine)
-					g_engine->pauseEngine(true);
-				GUI::MessageDialog alert(_("Do you really want to quit?"), _("Quit"), _("Cancel"));
-				result = _shouldQuit = (alert.runModal() == GUI::kMessageOK);
-				if (g_engine)
-					g_engine->pauseEngine(false);
-				_confirmExitDialogActive = false;
-			} else
-				_shouldQuit = true;
-
-			break;
 
 		default:
 			break;
 		}
 	} else {
 		// Check if event should be sent again (keydown)
-		if (_currentKeyDown.keycode != 0 && _keyRepeatTime < time) {
+		if (_currentKeyDown.keycode != Common::KEYCODE_INVALID && _keyRepeatTime < time) {
```

Comment:
I think there is (and has always been) an off-by-one in this time check, which should be `<=` or else this requires advancing 1ms past the next repeat time to repeat.

Reply:
Thanks, fixed